### ISSUE TITLE
plot_pareto_front crahses if a sampler is GridSampler

### ISSUE THREAD

**Issue by ilpoli** [optuna/optuna#4058]
### Expected behavior

plot_pareto_front should work with any sampler if there're multiple directions.

### Environment

- Optuna version: 3.0.2
- Python version: 3.7.14
- OS: Google Colab

### Error messages, stack traces, or logs

```shell
ValueError                                Traceback (most recent call last)
<ipython-input-67-dac422f460f1> in <module>
----> 9 fig = optuna.visualization.plot_pareto_front(study)
     10 fig.show()

3 frames
/usr/local/lib/python3.7/dist-packages/optuna/visualization/_pareto_front.py in plot_pareto_front(study, target_names, include_dominated_trials, axis_order, constraints_func, targets)
    129 
    130     info = _get_pareto_front_info(
--> 131         study, target_names, include_dominated_trials, axis_order, constraints_func, targets
    132     )
    133     return _get_pareto_front_plot(info)

/usr/local/lib/python3.7/dist-packages/optuna/visualization/_pareto_front.py in _get_pareto_front_info(study, target_names, include_dominated_trials, axis_order, constraints_func, targets)
    252         if include_dominated_trials:
    253             non_best_trials = _get_non_pareto_front_trials(
--> 254                 study.get_trials(deepcopy=False, states=(TrialState.COMPLETE,)), best_trials
    255             )
    256         else:

/usr/local/lib/python3.7/dist-packages/optuna/visualization/_pareto_front.py in _get_non_pareto_front_trials(trials, pareto_trials)
    362     non_pareto_trials = []
    363     for trial in trials:
--> 364         if trial not in pareto_trials:
    365             non_pareto_trials.append(trial)
    366     return non_pareto_trials

/usr/local/lib/python3.7/dist-packages/optuna/trial/_frozen.py in __eq__(self, other)
    181         if not isinstance(other, FrozenTrial):
    182             return NotImplemented
--> 183         return other.__dict__ == self.__dict__
    184 
    185     def __lt__(self, other: Any) -> bool:
```

### Steps to reproduce

```python
def objective(trial):
    a = trial.suggest_int("a", 100, 200, 10)
    b = trial.suggest_float("b", 0.5, 1, step= 0.1)
    
    return a + b


search_space = {
    "a": np.linspace(100, 200, 10+1), 
    "b": np.linspace(0.5, 1.0, 5+1),
}
study = optuna.create_study(
    directions=["minimize", "minimize"], 
    sampler=optuna.samplers.GridSampler(search_space))
study.optimize(objective, n_trials=1000, show_progress_bar=True)
```

### Additional context (optional)

I guess the issue is caused by an additional key_value in system attributes that is added by the grid sampler, 
```python
system_attrs={'search_space': OrderedDict([('b', array([0.5, 0.6, 0.7, 0.8, 0.9, 1. ])), ('a', array([100., 110., 120., 130., 140., 150., 160., 170., 180., 190., 200.]))]), 'grid_id': 1}
```
which can be compared with the default comparator
```shell
/usr/local/lib/python3.7/dist-packages/optuna/visualization/_pareto_front.py in _get_non_pareto_front_trials(trials, pareto_trials)
    362     non_pareto_trials = []
    363     for trial in trials:
--> 364         if trial not in pareto_trials:
```

**Comment by ilpoli** [comment]
Here's a workaround,

```
trials = study.trials
for trial in trials:
    del trial.system_attrs["search_space"]
fixed_study = optuna.create_study(directions = study.directions)
fixed_study.add_trials(study.trials)
fig = optuna.visualization.plot_pareto_front(fixed_study)
fig.show()
```

**Comment by toshihikoyanase** [comment]
Thank you for your bug report. I guess the `np.ndarray`s created by `np.linspace` prevent the comparison of trials.
 Since `search_space` of `GridSampler.__init__` expects `Sequence`, could you convert them to `list`s?

```python
search_space = {
    "a": np.linspace(100, 200, 10+1).tolist(), 
    "b": np.linspace(0.5, 1.0, 5+1).tolist(),
}
```

I didn't know but `np.ndarray` was not a `Sequence` type.

```python
import numpy as np
from typing import Sequence

a = np.linspace(100, 200, 10+1)
print(isinstance(a, Sequence))
print(isinstance(a.tolist(), Sequence))
```

After converting `np.ndarray` to `list`, I could plot the Pareto front as expected.

![image](https://user-images.githubusercontent.com/3255979/194984594-80a43552-3833-476b-9113-dc6294ca41df.png)


Please see [this notebook](https://colab.research.google.com/drive/1J9DfxQl3TseKGpKR1oMvJAKy2ONn0-fs?usp=sharing) for further details.

**Comment by toshihikoyanase** [comment]
I created a PR https://github.com/optuna/optuna/pull/4062 to fix this issue. Please feel free to add your comments. @ilpoli

**Comment by nzw0301** [comment]
can we close this issue?

**Comment by not522** [comment]
Let me close this issue since the bug has been fixed by #4062. Please feel free to reopen it if you still have issue.
